<a href="https://colab.research.google.com/github/naominicole27/Analitica-de-Negocios/blob/main/%C3%81rbol_de_Decisi%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Caso de Estudio**
Una entidad financiera (FINTECH) quiere implementar un modelo de árbol de decisión para mejorar la preaprobación de créditos de consumo de sus solicitantes de este tipo de créditos. Para este proceso vamos a utilizar las siguientes variables:
* Edad: Indica el npumero de ñaos que posee una persona, o el tiempo que usted lleva en el sistema financiero.
* Ingresos: Engloba todos los ingresos que recibe una persona, además si posee salario mensual (USD)
* Egresos:
* Monto (EAD):

0. Se procede con la carga de las librerías de trabajo.

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix


1. Se procede con la craga de los datos del trabajo.

In [ ]:
nxl='/content/1. Base de Datos - SolicitantesCrédito(USD).xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)

XD=XDB.iloc[:,[1,10,11,25]]
yd=XDB.iloc[:,32]
display(XDB)

,Unnamed: 0,Edad,Genero,Estado civil,Hijos,Perscargo,Nivel Renta,Estrato,Nivel de estudios,Tipo de contrato,...,Plazo,Cuota (COP),ConoCliente,Garantia,RespPat,MoraCome,PreApr,Prob.Default (PD),Score,LGD
0,1,38,NR,Casado,1,0,0,3,Bachiller,Ninguno,...,48,18.375890,Buenas Referencias,Sin Codeudor,Sin Patrimonio,Con Experiencia sin Mora,1,0.041659,582.478943,0.063702
1,2,51,NR,Separado,6,1,1,4,Técnico,Ninguno,...,36,5.062492,Buenas Referencias,Sin Codeudor,Prop.sin Hipoteca,Con Experiencia sin Mora,1,0.234951,479.839966,0.465636
2,3,37,Mujer,Casado,2,0,0,3,Tecnólogo,Indefinido,...,36,10.287688,Buenas Referencias,Con Codeudor(es),Sin Patrimonio,Con Experiencia sin Mora,0,0.170330,417.653107,0.290071
3,4,29,NR,Separado,1,0,0,2,Tecnólogo,Ninguno,...,24,15.458878,Buenas Referencias,Con Codeudor(es),Sin Patrimonio,Con Experiencia sin Mora,1,0.201659,316.791107,0.278389
4,5,42,Mujer,Casado,3,4,4,2,Especializacion,Ninguno,...,24,24.995180,Buenas Referencias,Con Codeudor(es),Sin Patrimonio,Con Experiencia sin Mora,1,0.140929,441.809570,0.203359
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5837,8372,48,Hombre,Separado,0,0,0,3,Tecnólogo,Fijo,...,36,27.043503,Buenas Referencias,Sin Codeudor,Prop.sin Hipoteca,Con Experiencia sin Mora,1,0.056132,550.980408,0.091441
5838,8374,31,Hombre,Soltero,2,0,0,5,Bachiller,Indefinido,...,48,25.579568,Buena Experiencia CFB,Sin Codeudor,Sin Patrimonio,Con Experiencia sin Mora,1,0.035120,659.471375,0.066962
5839,8375,38,Hombre,Separado,0,0,0,3,Bachiller,Labor Contratada,...,36,21.508792,Buenas Referencias,Sin Codeudor,Prop.sin Hipoteca,Con Experiencia sin Mora,1,0.101976,465.011536,0.165936
5840,8376,43,Mujer,Casado,1,2,2,2,Tecnólogo,Ninguno,...,36,11.047468,Buenas Referencias,Sin Codeudor,Sin Patrimonio,Con Experiencia sin Mora,1,0.156707,423.237305,0.247763


2. Se procede con la implementación del módulo árbol.

In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=4)
mar.fit(XD,yd) #Aquí el modelo busca la relación entrada - salida.

#¿Y qué fue lo que hizo el modelo?

ydp=mar.predict(XD)   #Esto es lo que pronosticó el modelo.

#Se construye la matrix de confusion.
cm=confusion_matrix(yd,ydp)

display(cm)
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

#Métricas de desempeño

Ex= (VP+VN)/len(XD)         #1. EXACTITUD: COMPORTAMIENTO GENERAL
print("La exactitud es:", Ex)

Sen=VP/(VP+FN)              #2. SENSIBILIDAD; COMO SE COMPORTA PRONOSTICANDO PREAPROBADOS.
print("La sensibilidad es:", Sen)

Spe=VN/(VN+FP)              #3. ESPECIFICIDAD: COMO SE COMPORTA FRENTE A LOS NEGATIVOS.
print("La especificidad es:", Spe)

Pre=VP/(VP+FP)              #4. PRECISION: COMO SE COMPORTA PRONOSTICANDO PREAPROBADOS
print("La precisión es:",Pre)

PreNeg=VN/(VN+FN)           #5. PRECISION NEGATIVA: COMO SE COMPORTA PRONOSTICANDO NEGATIVOS
print("La precisión negativa es:", PreNeg)


array([[2301,  658],
       [ 644, 2239]])

La exactitud es: 0.7771311194796303
La sensibilidad es: 0.7766215747485259
La especificidad es: 0.7776275768840825
La precisión es: 0.772868484639282
La precisión negativa es: 0.7813242784380305


**ANÁLISIS DE RESULTADOS**

De la base de datos se puede observar un total de 2959 solicitantes de crédito que poseen la categoría de prenegados, mientras que para la categoría de preaprobados la matriz cuenta con un total de 2883 créditos preaprobados.
De los 2959 datos el modelo pronosticó correctamente un total de 2301 datos, de los 2883 preaprobados el modelo pronosticó correctamente un total de 2239 datos.
El modelo logra identificar un 77% los créditos prenegados (2301/2959), mientras que el modelo logró identificar un 77.66% los créditos preaprobados (2239/2883).
Con respecto a las métricas podemos observar que el modelo logró una Exactitud  de 77.71%, lo que indica el buen comportameinto general del modelo frente a la clasificación de créditos en  las dos categorías de PreAprobación. Se destacan la especificidad y la precisión las cuales lograron valores en promedio por encima del 77%, lo que supera el límite inferiro definido para este tipo de modelos de clasificación el cual se ubica en el 75%.


De acuerdo con el árbol de decisión se destaca un nodo puro (10/0) el cual posee la siguiente regla de decisión, SI monto (>=322, ingr>=376, egr>= 685, ingr>= 644 (923/5) tendrá una probabilidad de aprobación del 100%. Es importante destacar que los nodos puros poseen un Gini = a 0, el modelo diferencia muy bien a los buenaspagas)

A pesar de la no existencia de más nodos puros,se destacan los nodos extremos o las reglas del negocio extremas. Se destaca una segunda regla que logra un porcentaje de aprobación 99% y la cual posee la siguiente estructura: Si monto <=322, ingr<=376, monto <=178, ing<= 232. Por su parte, el nodo derecho posee un porcentaje de negacion del 98%, en donde la regla se define: Monto>322, ingr> 896, Monto> 961, Ingr>1178, (7/576)

3. Despliegue del Árbol de Decisión

In [ ]:
from sklearn.tree import export_graphviz           #Exporta los datos a un gráfico
from pydotplus import graph_from_dot_data          #Es un graficador

vs=["Edad","Ingresos","Egresos","Monto"]           #Títulos del árbol
dot_data=export_graphviz(mar,feature_names=vs)     #Exportar de números a gráfico en pdf
graph=graph_from_dot_data(dot_data)                #Hacemos el gráfico
graph.write_png('Arbol.png')

NameError: name 'mar' is not defined